lightning/src/util/persist.rs

```diff
@@ -406,30 +406,62 @@ where
 		)?;
 		let mut res = Vec::with_capacity(monitor_list.len());
 		// for each monitor...
-		for maybe_monitor_name in monitor_list.into_iter().map(MonitorName::new) {
-			let monitor_name = maybe_monitor_name?;
-			// ...parse the monitor
-			let (block_hash, monitor) = self.read_monitor(&monitor_name)?;
-			// ...parse and apply the updates with an id higher than the monitor.
-			let mut current_update_id = monitor.get_latest_update_id();
-			loop {
-				current_update_id = match current_update_id.checked_add(1) {
-					Some(next_update_id) => next_update_id,
-					None => break,
-				};
-				let update_name = UpdateName::from(current_update_id);
-				let update = match self.read_monitor_update(&monitor_name, &update_name) {
-					Ok(update) => update,
-					Err(err) if err.kind() == io::ErrorKind::NotFound => {
-						// We can't find any more updates, so we are done.
-						break;
-					}
-					Err(err) => return Err(err),
-				};
+		for monitor_key in monitor_list {
+			// ...push the result into the return vec
+			res.push(self.read_channel_monitor_with_updates(
+				broadcaster,
+				&fee_estimator,
+				monitor_key,
+			)?)
+		}
+		Ok(res)
+	}
 
-				monitor
-					.update_monitor(&update, broadcaster, fee_estimator.clone(), &self.logger)
-					.map_err(|e| {
+	/// Read a single channel monitor, along with any stored updates for them.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is extremely important that your [`KVStore::read`] implementation uses the
+	/// [`io::ErrorKind::NotFound`] variant correctly: that is, when a file is not found, and _only_
+	/// in that circumstance (not when there is really a permissions error, for example). This is
+	/// because this function does not list updates. Instead, it lists all monitors, and then using
+	/// their stored `update_id`, synthesizes update storage keys, and tries them in sequence until
+	/// one is not found. All _other_ errors will be bubbled up in this function's [`io::Result`].
```

Comment:
Could we put this advise in one place and link to it rather than repeating it?

Reply:
The reason I repeated it is, both functions are now `pub` and are also pretty much exclusive ways of doing the same thing (that is, you'd either run the one that gets all monitors serially, or you would devise your own way of parallelizing one-at-a-time). So I'm concerned people could miss the warning.

I know it's a strong warning, but this is the only part of this design I feel is really dangerous. I'm open to ways to ensure it gets across without being so shouty about it.

Reply:
Oh, are you saying, maybe link to the other function?

Reply:
I put a summary and link in the "all" function docs.

Reply:
Hmm... I think this should go at the struct level as it involves a type parameter on the struct. Probably the "Reading channel state from storage" section, which can be linked from both methods, instead of the other way around.

I'm more inclined to move our docs in the direction of using more organized higher-level docs (linking back to it where necessary) so that readers have a natural entry point and don't need to jump around to get the full picture.